File: Companies/VMware/SlidingWindowMaximum.cpp
#include <iostream>
#include <vector>
#include <climits>
#include <deque>

using namespace std;

class SlidingWindowMaximum
{
public:
  vector<int> maxSlidingWindow(vector<int> &nums, int k)
  {
    vector<int> re{};
    deque<int> dp{};
    for (int i = 0; i < nums.size(); i++)
    {
      while (!dp.empty() && dp.front() < i - k + 1)
      {
        dp.pop_front();
      }
      while (!dp.empty() && nums[dp.back()] < nums[i])
      {
        dp.pop_back();
      }
      dp.push_back(i);
      if (i >= k - 1)
      {
        re.push_back(nums[dp.front()]);
      }
    }
    return re;
  }
};

File: Companies/VMware/SlidingWindowMaximum.java
package Companies.VMware;

public class SlidingWindowMaximum {
    public int[] maxSlidingWindow(int[] nums, int k) {
        if (nums.length == 0) {
            return new int[]{};
        }
        int[] re = new int[nums.length-k+1];
        int i = 0, maxp = 0, max = Integer.MIN_VALUE;
        while (i < k) {
            if (nums[i] >= max) {
                max = nums[i];
                maxp = i;
            }
            i++;
        }
        re[0] = max;
        while (i < nums.length) {
            if (nums[i] >= max) {
                max = nums[i];
                maxp = i;
            } else {
                if (maxp < i-k+1) {
                    max = nums[i];
                    maxp = i;
                    for (int j = i-1; j >= i-k+1; j--) {
                        if (nums[j] >= max) {
                            max = nums[j];
                            maxp = j;
                        }
                    }
                }
            }
            re[i-k+1] = max;
            i++;
        }
        return re;
    }
}
